Adding description of unknown property name handling (#162)
Materials-Consortia/OPTiMaDe: optimade.md

  For each entry type, the relationships with entries of that type can then be encoded in a field with the name of the entry type, which are to contain a list of the IDs of the referenced entries alongside the respective human-readable description of the relationships. 
  It is the intent that future versions of this standard uphold the viability of this encoding by not standardizing property names that overlap with the entry type names.

## <a name="h.3.6">3.6. Handling unknown property names</a>

> If a property name is not recognised by the queried end-point, an

Reviewer: ```suggestion
If a property name is not recognized by the queried endpoint, an
```
(endpoint is written like this throughout the rest of the text)

error diagnosis would be appropriate. However, we also want to

> formulate a single query suitable for multiple databases, even when

Reviewer: ```suggestion
support the formulation of a single query suitable for multiple databases, even when
```

> querying database-sepcific properties. For example, we would like to

Reviewer: ```suggestion
querying database-specific properties. For example, we would like to
```

> construct the following query:

Reviewer: ```suggestion
support the following query:
```

Author: Maybe better: "support construction of the following query"?

Reviewer: Can you try explain what are you aiming to convey with the extra word "construct" here? 

I replaced it with "support" because formally the API specification doesn't deal with the "construction" of anything, only what is a valid and invalid API request and API response.

Also, while we are discussing the formulations here: I don't find it very clear who 'we' refers to in this paragraph, so I'd also suggest rephrasing those.

Author: @rartino : your suggestion "support" does not fit into the current text: it is the backend that "supports" the query; but the client (as is implied in the text) "constructs" the query (in a very general sense – composes the query from characters). I leave "construct" at the moment for the lack of better word; "maybe "compose a query" would be a better alternative, but definitely not the "support".

Author: > Also, while we are discussing the formulations here: I don't find it very clear who 'we' refers to in this paragraph, so I'd also suggest rephrasing those.

Changing "we" to "an OPTiMaDe client".

> `filter=_aida_bandgap<2.0 OR _mp_bandgap<2.5`

Reviewer: I'd like this anonymized. In my opinion the specification should be neutral in its 'endorsement' of specific databases, even in examples. E.g.,
```suggestion
`filter=_exmpl_bandgap<2.0 OR _exmpl2_bandgap<2.5`
```
with some additional text defining that `_exmpl_` and `_expl2_`  are supposed to be a database-specific prefix of two different databases.

Also, if this text anyway remains, the `_aiida_` prefix is misspelled.

and then send it query to AiiDA-backed and MP-backed OPTiMaDe
endpoints.

> For these queries to succeed, the following behavior is suggested:

Reviewer 2: There are a few other minor typos or minor things that I would change, but I start with the main question: do we want to SUGGEST this, or enforce this? Because below we then say MUST and I think this is now confusing.

Author: The formal conformance level is MUST; 'suggested' was a milder expression for humans ;)

Should we change 'suggested' -> 'required'?

Reviewer 2: even better, use proper RFC e.g. `MUST`

Reviewer 3: I agree that "suggested" should be proper RFC, e.g. one of `MAY`, `SHOULD`, `MUST`.

Author: 1b10d49 Rewording the phrase 'behavior is suggested' using the RFC2119 "MUST" keyword.

> * if a database receives a query filter with universal Optimade

Reviewer: ```suggestion
* if a database receives a query filter with universal OPTiMaDe
```

>   properties (i.e. defined without a database prefix, and deemed

Reviewer: ```suggestion
  properties (i.e., defined without a database prefix, and deemed
```

  universal for all databases), or a property with the database's

>   private prefix (e.g. cod for the COD database), then the database

Reviewer: ```suggestion
  private prefix (e.g. `_exmpl_`), then the API implementation
```

  MUST check if the provided property names are known, and MUST return

>   an appropriate error code if they are not known the database.

Reviewer: ```suggestion
  an appropriate error code if they are not known to the API implementation.
```

Author: If we change to "API implementation" then we will probably have to change it elsewhere in the suggested text as well; but is it really the API implementation that decides on the names or the database? It is actually the interaction of both; without the database providing a property API an API implemenatation can not support the name; but API can also fail to support some property name even if the database has corresponding data...

Reviewer: We define what API Implemenation means in term definitions, and elsewhere in the specification text we use this term with the meaning you are looking for here, i.e., when it is the "interaction" between the API implementation + the database backend.

Author: OK, I'll have a careful look at the current wording and adapt the text accordingly.

Author: See commits https://github.com/sauliusg/OPTiMaDe/commit/7d8d3cc54a92631d2ea94f39e4e06d451f2a0fb1, https://github.com/sauliusg/OPTiMaDe/commit/b9630a5dc2af680f0d989b3b5fb3bf4961a585d3 and https://github.com/sauliusg/OPTiMaDe/commit/906d27aec803dd3cd89ea4322b11bb7b9e1a8ec3 for a suggested wording.

* if a database received a query filter with a property from an
  unknown database, the end-point MUST behave as if that property is

>   unknown, i.e. is if it has the value null. If the database prefix of

Reviewer: (For rationale, see separate comment)

In my opinion 
> If the database prefix of the unknown property is similar to that of a known database, the response MAY issue a corresponding warning.

should be replaced with

> The API implementation SHOULD issue warnings for every unrecognized property and specify that they have been handled as unknown.

Author: After some though I realise that issuing warning on every unrecognised property (even if we know that it is not "our" property, i.e. we recognise another database's prefix) would **always** issue warnings on mixed-database queries. These expected warnings will then obscure the "real" (mistyped) ones; user (including myself) will tend to ignore warnings if they are too numerous and mostly uninformative. (Let's not shout "wolf!" too often). This consideration would point to issuing warnings only when something indeed might be wrong.

What about such behaviour:

- if a database prefix is **recognised** by an API implementation as belonging to a different existing database, no warning is issued, or different kind of warning is issued (e.g. "property 'xyz' for database 'example9' was assumed to be `null`"); this warning could have MAY or SHOULD conformance level;

- if a database prefix is **not** recognised, we issue a warning (e.g. "database 'exxxxample' was not recognised"); conformance level for this warning would be SHOULD.

Is that OK? If you are fine with it, I'll change the text accordingly.

Reviewer: I see your rationale, and can accept this solution. I assume @giovannipizzi prefers this over my suggestion as well, since it will give less warnings.

Do we have a facility for INFO level messages? I may want to still trigger those in this case (but, completely optionally for my implementation; I wouldn't want to mandate INFO level messaging.)

Author: The solution is, I believe, implemented in the latest commits of this PR.

Author: > Do we have a facility for INFO level messages? I may want to still trigger those in this case (but, completely optionally for my implementation; I wouldn't want to mandate INFO level messaging.)

I'm not sure if the current implementation specifies message levels formally, but it would be helpful to have ERROR, WARNING and NOTE/INFO level messages. The distinction would be as follows:

- ERROR would mean that the API implementation could not server the request. No data is provided, and the computations can not reasonably continue on the response.

- WARNING would mean that the response contains valid data, and formally the computations can continue, but the result is probably not what the client expects. For example, the query might contain a misspelled foreign database property and return 0 entries as a response (which is valid), but there is a hight probability that the property name was mistyped and the client expects query on a different property;

- NOTE/INFO would mean that the query ran as expected by the API implementation, and the message just states what has been done to make the situation explicit in cases where it otherwise might be ambiguous.

  the unknown property is similar to that of a known database, the

>   response MAY issue a corresponding warning.

Reviewer 2: Why do we also want to issue a warning? I understand if this is possible mistype of *your* db (let's say you find `_cdo_` and you are `_cod_` and you don't have `_cdo_` in the list of optimade providers, but in general, is this useful?

Author: I think warnings are very useful in this case: imagine you query the COD, and instead of `_cod_deposition_date > "2019-01-01"` you mistype `_cdo_deposition_date > "2019-01-01"`. The OPTiMaDe back end assumes that `_cdo_deposition_date` is always `null`, thus the comparison is always `false`, and you end up with no structures instead of a required ones. This can be very confusing, especially of you query has another ORed components that **do** return some structures – you would en up in incorrect selection, but this would be hard to detect, and the typo is difficult to notice as well.

A warning would draw your attention to a possible mistake; it is supposed to detect typos similar to a known database names ('cod' -> 'cdo' would most probably be considered "similar"). Still, the formal similarity level is not specified, and the implementation of similar string identification might be tricky, so I left this requirement at the "MAY" level, although "SHOULD" could be discussed since the feature is very useful (we do not want OPTiMaDe to be used to retrieve garbage data, do we?)

Reviewer 2: I agree - but my point is that you need to check only for your own name and similar ones - e.g. if COD finds a `_aidia` instead of `_aiida` it shouldn't complain - it could just be a different, recent/new DB that it does not know. If instead you find `_cdo` you should warn. 

By reducing the warnings, we make sure these are not very often printed, that will in turn mean that the user will start not even reading them

Reviewer: I'm in favor of that implementations SHOULD issue a warning for any unknown property name encountered that does not trigger an error.

Sending queries that include other databases' fields is going to be something quite unusual anyway, where a user should know that what they are doing is "odd" and therefore they shoulde expect warnings. I don't think this is an issue of over-using the warning system.

Reviewer 2: Ok. But then should we then just trigger a warning anyway, independent of string "closeness" or "similarity" to known strings?

Reviewer: I'd say no to string closeness comparisons.

To clarify, in my opinion:
* Global properties and properties with the providers own prefix, which the implementation *can* verify as non-existing should generate errors, just as described in this PR.
* Properties with a different db prefix, which the implementation thus *cannot* verify as existing or non-existing, are to be handled in filters as unknowns as described in this PR, but SHOULD always generate warnings about being handled this way. 

When *not* trying to do the type of query this PR tries to handle, these warnings signal that the user has misspelled the database's own prefix (perhaps badly so). They are thus highly relevant.

In queries of the type this PR tries to handle, these warnings will be somewhat useful in confirming that the databases does what the user intends. E.g. *"Warning: unrecognized property \_aida\_bandgap handled as unknown"*. I would very much expect that warning if I sent the example query given in the text of this PR to the MP OPTiMaDe implementation.

Author: > Sending queries that include other databases' fields is going to be something quite unusual anyway, where a user should know that what they are doing is "odd" and therefore they should expect warnings. I don't think this is an issue of over-using the warning system.

I agree with @rartino that the database SHOULD issue a warning (or a note), but for somewhat different reason :).

I think the situation when you send requests with multiple database prefixes is quite normal and there is nothing "odd" about it, but the warnings (or even note-level messages) will not obscure other ones since one can (and should ;) filter them out. These messages just explicitly state what has happened in a backend, making the situation explicit.
